packages/itwinui-react/src/core/ThemeProvider/ThemeProvider.tsx

```diff
@@ -150,55 +163,45 @@ export const ThemeProvider = React.forwardRef((props, forwardedRef) => {
   themeOptions.highContrast ??=
     themeProp === 'inherit' ? parent.highContrast : undefined;
 
-  /**
-   * We will portal our portal container into `portalContainer` prop (if specified),
-   * or inherit `portalContainer` from context (if also inheriting theme).
-   */
-  const portaledPortalContainer =
-    portalContainerProp ||
-    (themeProp === 'inherit' ? parent.context?.portalContainer : undefined);
-
-  const [portalContainer, setPortalContainer] = useControlledState(
-    null,
-    portaledPortalContainer,
-  );
+  const [portalContainerFromParent] = useScopedAtom(portalContainerAtom);
 
   const contextValue = React.useMemo(
-    () => ({ theme, themeOptions, portalContainer }),
+    () => ({ theme, themeOptions }),
     // we do include all dependencies below, but we want to stringify the objects as they could be different on each render
     // eslint-disable-next-line react-hooks/exhaustive-deps
-    [theme, JSON.stringify(themeOptions), portalContainer],
+    [theme, JSON.stringify(themeOptions)],
   );
 
   return (
-    <HydrationProvider>
-      <ThemeContext.Provider value={contextValue}>
-        {includeCss && rootElement ? (
-          <FallbackStyles root={rootElement} />
-        ) : null}
-
-        <Root
-          theme={theme}
-          themeOptions={themeOptions}
-          ref={useMergedRefs(forwardedRef, setRootElement)}
-          {...rest}
-        >
+    <ScopeProvider>
+      <HydrationProvider>
+        <ThemeContext.Provider value={contextValue}>
           <ToastProvider>
-            {children}
-
-            {portaledPortalContainer ? (
-              ReactDOM.createPortal(<Toaster />, portaledPortalContainer)
-            ) : (
-              <div ref={setPortalContainer} style={{ display: 'contents' }}>
-                <Toaster />
-              </div>
-            )}
+            {includeCss && rootElement ? (
+              <FallbackStyles root={rootElement} />
+            ) : null}
+
+            <Root
+              theme={theme}
+              themeOptions={themeOptions}
+              ref={useMergedRefs(forwardedRef, setRootElement)}
+              {...rest}
+            >
+              {children}
+
+              <PortalContainer
+                portalContainerProp={portalContainerProp}
+                portalContainerFromParent={portalContainerFromParent}
+                isInheritingTheme={themeProp === 'inherit'}
+              />
+            </Root>
           </ToastProvider>
-        </Root>
-      </ThemeContext.Provider>
-    </HydrationProvider>
+        </ThemeContext.Provider>
+      </HydrationProvider>
+    </ScopeProvider>
   );
 }) as PolymorphicForwardRefComponent<'div', ThemeProviderOwnProps>;
+ThemeProvider.displayName = 'ThemeProvider';
```

Comment:
nit: Was this necessary for some debugging? Is the `displayName` not set automatically?

Reply:
Yes, it shows up as `ForwardRef` in React dev tools